hdeps/compatibility.py

```diff
@@ -19,55 +18,75 @@ def find_best_compatible_version(
     project: Project,
     req: Requirement,
     env_markers: EnvironmentMarkers,
-    already_chosen: Optional[LooseVersion],
+    already_chosen: Optional[Version],
     current_version_callback: VersionCallback,
-) -> LooseVersion:
+) -> Version:
     # Handle requires_python first, so we can produce a better error message
     # when there are no version-compatible candidates (before we even get to
     # req.specifier)
     python_version_str = env_markers.python_full_version
     assert python_version_str is not None
     python_version = Version(python_version_str)
-    possible: List[LooseVersion] = []
 
     requires_python_cache: Dict[str, bool] = {}
 
-    with kev("filter requires_python"):
-        for v, pv in project.versions.items():
-            try:
-                if pv.requires_python:
-                    result = requires_python_cache.get(pv.requires_python)
-                    if result is None:
-                        ss = SpecifierSet(pv.requires_python)
-                        result = python_version in ss
-                        requires_python_cache[pv.requires_python] = result
-                    if not result:
-                        continue
-            except InvalidSpecifier as e:
-                LOG.debug("Ignore %s==%s because %r", project.name, v, e)
-                continue
-            possible.append(v)
+    def requires_python_match(v):
+        pv = project.versions[v]
+        if not pv.requires_python:
+            return True
+        if verdict := requires_python_cache.get(pv.requires_python) is not None:
+            return verdict
+        try:
+            specifier_set = SpecifierSet(pv.requires_python)
+            verdict = python_version in specifier_set
+        except InvalidSpecifier as e:
+            LOG.debug(
+                "Ignore %s==%s has invalid requires_python %r but including anway",
+                project.name,
+                v,
+                e,
+            )
+            verdict = True
+        requires_python_cache[pv.requires_python] = verdict
+        return verdict
+
+    possible: List[Version] = []
+
+    specifier_matched = False
+    with kev("reverse"):
+        rev = reversed(project.versions.keys())
+    with kev("initial filter"):
+        for version in req.specifier.filter(rev):
+            specifier_matched = True
+            if requires_python_match(version):
+                # Only keep the first one
+                possible.append(version)
+                break
```

Comment:
Would this be the same as this?
```
possible = any(v for v in req.specifier.filter(rev) if requires_python_match(v))
```

Reply:
Oops, not quite, more like a `possible.extend()` but then it's less readable

Reply:
Yeah, and I'm pretty sure the magic around prerelease versions got lost.  It's about this point that it became a draft PR, clearly needs more testing.